libs/hwdrivers/src/rplidar/src/rplidar_driver.cpp

```diff
@@ -298,6 +411,116 @@ u_result RPlidarDriverSerialImpl::_cacheScanData()
+            currentAngle_raw_q16 += angleInc_q16;
+
+            for (int cpos = 0; cpos < 2; ++cpos) {
+
+                if (angle_q6[cpos] < 0) angle_q6[cpos] += (360<<6);
+                if (angle_q6[cpos] >= (360<<6)) angle_q6[cpos] -= (360<<6);
+
+                rplidar_response_measurement_node_t node;
+
+                node.sync_quality = (syncBit[cpos] | ((!syncBit[cpos]) << 1));
+                if (dist_q2[cpos]) node.sync_quality |= (0x2F << RPLIDAR_RESP_MEASUREMENT_QUALITY_SHIFT);
+
+                node.angle_q6_checkbit = (1 | (angle_q6[cpos]<<1));
+                node.distance_q2 = dist_q2[cpos];
+
+                nodebuffer[nodeCount++] = node;
+             }
+
+        }
+    }
+
+    _cached_previous_capsuledata = capsule;
+    _is_previous_capsuledataRdy = true;
+}
+
+
+u_result RPlidarDriverSerialImpl::_cacheCapsuledScanData()
+{
+    rplidar_response_capsule_measurement_nodes_t    capsule_node;
+    rplidar_response_measurement_node_t      local_buf[128];
+    size_t                                   count = 128;
+    rplidar_response_measurement_node_t      local_scan[MAX_SCAN_NODES];
+    size_t                                   scan_count = 0;
+    u_result                                 ans;
+    memset(local_scan, 0, sizeof(local_scan));
+
+    _waitCapsuledNode(capsule_node); // // always discard the first data since it may be incomplete
+
+    while(_isScanning)
+    {
+        if (IS_FAIL(ans=_waitCapsuledNode(capsule_node))) {
+            if (ans != RESULT_OPERATION_TIMEOUT && ans != RESULT_INVALID_DATA) {
+                _isScanning = false;
+                return RESULT_OPERATION_FAIL;
+            } else {
+                // current data is invalid, do not use it.
+                continue;
+            }
+        }
+
+        _capsuleToNormal(capsule_node, local_buf, count);
+
+        for (size_t pos = 0; pos < count; ++pos)
+        {
+            if (local_buf[pos].sync_quality & RPLIDAR_RESP_MEASUREMENT_SYNCBIT)
+            {
+                // only publish the data when it contains a full 360 degree scan 
+                
+                if ((local_scan[0].sync_quality & RPLIDAR_RESP_MEASUREMENT_SYNCBIT)) {
+                    _lock.lock();
+                    memcpy(_cached_scan_node_buf, local_scan, scan_count*sizeof(rplidar_response_measurement_node_t));
+                    _cached_scan_node_count = scan_count;
+                    _dataEvt.set();
+                    _lock.unlock();
+                }
+                scan_count = 0;
+            }
+            local_scan[scan_count++] = local_buf[pos];
+            if (scan_count == _countof(local_scan)) scan_count-=1; // prevent overflow
+        }
+    }
+    _isScanning = false;
+
+    return RESULT_OK;
+}
+
 u_result RPlidarDriverSerialImpl::grabScanData(rplidar_response_measurement_node_t * nodebuffer, size_t & count, _u32 timeout)
 {
     switch (_dataEvt.wait(timeout))
```

Comment:
Wait has type of unsigned long. The right thing to do is fix rplidar. wait should return an enum.
This shouldn't be anonymous
```
enum
{
  EVENT_OK = 1,
  EVENT_TIMEOUT = -1,
  EVENT_FAILED = 0,
};
```
This should return an enum
```
unsigned long wait( unsigned long timeout = 0xFFFFFFFF )
```
The quick solution is explicitly cast each one of those case statements into a unsigned long.